### DISCUSSION TITLE
Do all rtcpeerconnections share the same udp handle/socket?

### DISCUSSION THREAD

**Discussion by unicomp21** [paullouisageneau/libdatachannel#1248] [General]
I'm told this is the case in pion (ie optimization), but am not sure if it's true for libdatachannel?

**Comment by paullouisageneau** [comment]
This is not the case in pion by default, since by specification each peer connection must use a different socket. However, libdatachannel also has an optional UDP mux mode that runs all peer connections over a single UDP socket. You can enable it with `enableIceUdpMux` in configuration. The mode should only be enabled on servers, as you might encounter connection issues if both sides have the mode enabled.

**Reply by unicomp21** [reply]
Thanks!  Why would there be issues if both ends have it enabled?  I'd love to have it enabled for client/server when doing p2p between servers w/ the datachannel(s) running in lossy mode?

**Reply by paullouisageneau** [reply]
Yes, it can be an issue in some situations if enabled on both ends. You won't be able to open multiple peer connections between the two hosts, which can prevent successful reconnections. 

This is a per-connection setting and the library supports both modes at the same time, I'd advise for a server to set `enableIceUdpMux = true` when connecting to a client, and `enableIceUdpMux = false` when connecting to a server, this way only connections to clients will share a single socket.

**Reply by unicomp21** [reply]
Hypothetical.  With a code change (ie added correlation id), could all datachannel connections within a process share the same udp socket?  My reason for asking, this could be useful for kernel-bypass stuff.  Would you entertain a PR if there was sufficient test coverage?

**Reply by paullouisageneau** [reply]
No, I'm afraid this is impossible due to how peer connections are demultiplexed at lower level.

Note that you can enable UDP mux on both modes, only you have to ensure there can be only a single connection between two hosts, and you must pay attention that the current connection bas been closed on both sides before attempting to reconnect.

**Reply by unicomp21** [reply]
When enabled on both ends, how do things break?  What are the side effect(s) seen by the app level?  Could I improvise by adding correlation id(s) at the app level?

Apologies for all the questions, I'm a bit excited/caffeinated, lol.  This is an amazing project!

**Reply by paullouisageneau** [reply]
Both connections would conflict and experience connectivity issues since endpoints can't reliably map incoming packets to each connection (ICE relies on address and port to demultiplex user trafic). There is nothing you can do to fix it at the app level.

**Reply by unicomp21** [reply]
Got it, thanks for teaching me!  Didn't realize it wasn't a 4-tuple (ie address/port for both ends to make key/lookup).
